cmd/contour/serve.go

```diff
@@ -303,9 +315,25 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
 	eh.Metrics = metrics
 	eh.CacheHandler.Metrics = metrics
 
+	synced := make([]cache.InformerSynced, 0, len(informers))
+	for _, inf := range informers {
+		synced = append(synced, inf.HasSynced)
+	}
+
 	// step 13. create grpc handler and register with workgroup.
 	g.Add(func(stop <-chan struct{}) error {
 		log := log.WithField("context", "grpc")
+
+		log.Printf("waiting for informer caches")
+		cache.WaitForCacheSync(stop, synced...)
+		log.Printf("informer caches synced")
+
+		// Wait for cache to process
+		timeout := 15 * time.Second // Default timeout if reached code still progresses
+		if !eh.WaitForCacheProcess(timeout) {
+			log.Printf("!! timed out waiting %v for cache to process !!", timeout)
```

Comment:
I think there are 3 basic cases to handle:
1. The cache has not been built yet
2. The cache is still being rebuilt
3. The cache has been built (at least once)

`WaitForCacheProcess` is addressing (2), though the ordering of operations in the event handler means that it can succeed before the DAG has actually been sent to the cache. IIUC, we want to guarantee that we always fall into case (3). The event handler Sequence also doesn't have the semantics we want for this.

One option to consider is an event handler method that explicitly forces a cache rebuild. You'd have to call `UpdateNow()` and then need to have a way to block until the update is done. If the sequence counter was atomic, you could sleep wait until the sequence changed. It feels a bit hack but it would work :)